### PR TITLE
Improve inpost metabox design

### PR DESCRIPTION
**Pull request by danielpost**
Hey there,

I thought the posts and pages metabox was a little messy design-wise, so I've made some small improvements (similar to ACF) to improve the flow and make it a little easier to use.

I haven't changed the tsf.min.css file yet. Would love to hear your thoughts!

screen shot 2017-03-09 at 17 22 27

### CODE REVIEW

**Comment by sybrew** [comment]
😮 I was just about to begin with adding tabs (see #62), but you just blew my mind with that!

I'll get back to you ASAP because this is just awesome!

**Comment by danielpost** [comment]
Glad to hear it! I really just stole ACF's design and implemented it. Could use the same for tabs too (see screenshot). Let me know if I can do anything!

screen shot 2017-03-09 at 17 43 04

File: lib/css/tsf.css [hunk 1]
```diff
@@ -538,6 +538,60 @@ body.js .tsf-metaboxes .tsf-tabs-content.tsf-active-tab-content {
 	display: block;
 }
 
+#tsf-inpost-box .inside {
+	margin: 0;
+	padding: 0;
+}
+
+#tsf-inpost-box .inside:after,
```
Comment (sybrew) [inline lib/css/tsf.css]: Use class "clearfix" instead?

File: lib/css/tsf.css [hunk 2]
```diff
@@ -538,6 +538,60 @@ body.js .tsf-metaboxes .tsf-tabs-content.tsf-active-tab-content {
...
+	padding: 15px 12px;
+	-webkit-box-sizing: border-box;
+    -moz-box-sizing: border-box;
+    box-sizing: border-box;
+}
+
+.tsf-singular-metabox-label {
+	width: 24%;
+	position: relative;
+	z-index: 2;
```
Comment (sybrew) [inline lib/css/tsf.css]: z-index adds code-debt, is this necessary?

File: lib/css/tsf.css [hunk 3]
```diff
@@ -538,6 +538,60 @@ body.js .tsf-metaboxes .tsf-tabs-content.tsf-active-tab-content {
...
+}
+
+.tsf-singular-metabox-label .tsf-counter {
+	display: block;
```
Comment (sybrew) [inline lib/css/tsf.css]: This makes the clickable field larger than wished for. On this class there's a JS-listener that allows switching of types.
Can be solved by wrapping another div around it with these properties. ".tsf-counter" has to be preserved for JS.

Comment (sybrew) [inline lib/css/tsf.css]: i.e. ".tsf-counter-wrap"

**Comment by sybrew** [comment]
ACF design is according to WordPress spec, so there's nothing wrong with using it 😄. 
I do like the tab lay-out, it was already to be planned to be on the left. 

I'll take on the tab part after the original Pull has been implemented because the tabs requires filters, dashicons, no-JS layered compat, etc.

File: inc/views/inpost/seo-settings.php [hunk 4]
```diff
@@ -134,135 +134,162 @@
...
-				<input class="large-text" type="text" name="autodescription[_genesis_title]" id="autodescription_title" placeholder="<?php echo esc_attr( $doctitle_placeholder ); ?>" value="<?php echo esc_attr( $this->get_custom_field( '_genesis_title' ) ); ?>" />
-				<span id="tsf-title-offset" class="hide-if-no-js"></span><span id="tsf-title-placeholder" class="hide-if-no-js"></span>
+		<div class="tsf-singular-metabox">
+			<div class="tsf-singular-metabox-label">
+				<label for="autodescription_title"><strong><?php printf( esc_html__( 'Custom %s Title', 'autodescription' ), esc_html( $type ) ); ?></strong>
+					<a href="<?php echo esc_url( 'https://support.google.com/webmasters/answer/35624?hl=' . $language . '#3' ); ?>" target="_blank" title="<?php esc_attr_e( 'Recommended Length: 50 to 55 characters', 'autodescription' ); ?>">[?]</a>
+					<span class="description tsf-counter">
+						<?php printf( esc_html__( 'Characters Used: %s', 'autodescription' ), '<span id="autodescription_title_chars">' . (int) mb_strlen( $tit_len_parsed ) . '</span>' ); ?>
+						<span class="hide-if-no-js tsf-ajax"></span>
+					</span>
```
Comment (sybrew) [inline inc/views/inpost/seo-settings.php]: Move this span block just below the label tag, it's weird now when clicking on it (the label on-click affects focusing on the input container).

File: inc/views/inpost/seo-settings.php [hunk 5]
```diff
@@ -134,135 +134,162 @@
...
+
+		<div class="tsf-singular-metabox">
+			<div class="tsf-singular-metabox-label">
+				<label for="autodescription_description">
+					<strong><?php printf( esc_html__( 'Custom %s Description', 'autodescription' ), esc_html( $type ) ); ?></strong>
+					<a href="<?php echo esc_url( 'https://support.google.com/webmasters/answer/35624?hl=' . $language . '#1' ); ?>" target="_blank" title="<?php esc_attr_e( 'Recommended Length: 145 to 155 characters', 'autodescription' ); ?>">[?]</a>
+					<span class="description tsf-counter">
+						<?php printf( esc_html__( 'Characters Used: %s', 'autodescription' ), '<span id="autodescription_description_chars">' . (int) mb_strlen( $desc_len_parsed ) . '</span>' ); ?>
+						<span class="hide-if-no-js tsf-ajax"></span>
+					</span>
```
Comment (sybrew) [inline inc/views/inpost/seo-settings.php]: Move this span block just below the label tag, it's weird now when clicking on it (the label on-click affects focusing on the input container).

File: lib/css/tsf.css [hunk 6]
```diff
@@ -538,6 +538,60 @@ body.js .tsf-metaboxes .tsf-tabs-content.tsf-active-tab-content {
...
+	position: absolute;
+	content: '';
+	top: 0;
+	left: 0;
+	bottom: 0;
+	width: 24%;
+	background-color: #f9f9f9;
+	border-right: 1px solid #e1e1e1;
+}
```
Comment (sybrew) [inline lib/css/tsf.css]: This block is a hack. I'll fix this later 😄

File: lib/css/tsf.css (hunk 3 above)
Comment (sybrew) [inline lib/css/tsf.css]: I'll do this, it will create a couple of browser bugs if you follow my request as-is.

File: lib/css/tsf.css (hunk 2 above)
Comment (sybrew) [inline lib/css/tsf.css]: Can be removed 😄

File: lib/css/tsf.css (hunk 1 above)
Comment (sybrew) [inline lib/css/tsf.css]: Nevermind, such change will break the layout.

**Comment by danielpost** [comment]
Great, thank you! Let me know if you need help with anything.

**Comment by sybrew** [comment]
Sure thing! Just a fair warning: I'm going to convert the whole thing to CSS3 flexbox 😄, so it will automatically fix the layout on various screens. This was something I already had in mind a long time ago :)

Here you can find the base wrapper classes: [tsfem.css](https://github.com/sybrew/The-SEO-Framework-Extension-Manager/blob/master/lib/css/tsfem.css#L37-L102) 
However, I'm going to start without such CSS class framework for now.

**Comment by sybrew** [comment]
@danielpost 

A or B?

A:
image

B:
image

**Comment by danielpost** [comment]
@sybrew 

I much prefer A, more consistent. I'm not sure about aligning the counter to the bottom, it would make sense if all fields had more space in between the label and the counter, but right now it makes it look less consistent imo.

Also, another idea: Instead of "Custom Social Image URL" say "Custom Social Media Image" and show a thumbnail instead of the URL.

**Comment by sybrew** [comment]
A it is 😄 

The counter to the bottom issue was an unfortunate side-effect I still have to solve.

Adding a thumbnail was something I already had in mind, but I also wanted to do that for no-JS. So that's for a later time. I also want to fetch the image on manual input and show if it's not a 404 link through AJAX.

Using "Media" could be confusing; because, "Media" is also a sidebar menu item and users could link those in their mind. 
Although this happens when you hit the "Select Image" button, there also has alternative behavior where you manually input an URL.

**Comment by sybrew** [comment]
@danielpost 
I know it's late, but I promise this is the (second-to) final item 😄 

A or B? Notice the centering of the labels.

A:
image

B:
image

Cheers 😄

**Comment by danielpost** [comment]
@sybrew

Hmm, this is a tough one. I like both but I'd still pick A here, also in case you ever get longer meta boxes with more content.

**Comment by sybrew** [comment]
When they get longer, the centering will flow with it. 

I also centered the right side now, I think that was making it tough 😄:
image

Still A?

**Comment by LeBaux** [comment]
I kinda feel that is maybe matter of taste rather than crucial design decision. Nevertheless, it is amazing to see that some other developers are taking interest in the project... it is breaking my heart to see 1 developer going 110% in one project with no support, so it is definitely welcomed addition from you @danielpost and I can't speak for @sybrew, but another person looking at design/CSS part of things is always welcomed addition.

If you ask me, it does not matter at the slightest. Honestly, I think the difference between A/B is marginal and it brings exactly same value to the end customer (imho, don't linch me).

**Comment by ramon-villain** [comment]
I liked the B version, and also liked the "label" idea for counting characters!

**Comment by danielpost** [comment]
@sybrew I still think A is better. In your screenshot, if you scroll down you can see the top of the textarea but not the label yet, so you won't immediately know what input it is. UX-wise A makes more sense. But I do agree it's very minor.

@LeBaux Thanks for the kind words! I know most users won't care, but it's up to developers to make sure it's something that users won't even notice or have to think about. Good UX is not one big decision but rather a huge collection of tiny decisions, every little thing matters.

**Comment by sybrew** [comment]
I definitely prefer the right-side to be centered.... the Title input looks "off" when top-aligned because the label is bigger.
Because of this, naturally the left side should also be centered.
However, I do agree with Daniel's comment regarding UI vs UX, where I implied it's more important to have a good user flow than a pretty layout for administrative purposes.

That said, it's still a mixed bag of opinions.
I'm going to center it for now, because I like it more 😅 

**Making it optional?***
Shall I make a triple-click listener that will save an option (like the counter type) to the user's options array through AJAX that will shift it from center aligned to top aligned? The framework for said options is already in place so it adds only a nanosecond or so of load time.

Or, a mini-button next to the collapse arrow of the metabox that has an equals sign (middle) vs a tie (top)? i.e. `=` vs  `⁀` or `-` vs ` ͞`?

`†` vs `‡` is out of the question because the first symbol can remind one of death 😓.

**Comment by danielpost** [comment]
Fair enough. I'd stick with centered then, the input fields are short enough that it won't be an issue then. I definitely wouldn't make it optional, it's a very minor thing and it would only confuse users (also see [decisions, not options](https://tommcfarlin.com/wordpress-philosophy-decisions-not-options/)).

**Comment by sybrew** [comment]
It won't necessarily confuse users when there's a direct effect. It's like switching the "justify-content" option off in the browser console. I'll push the commit where you can test this. Obviously, when implemented, it should animate.

These are what you should turn off:
image

Note that you require `define( 'SCRIPT_DEBUG', true );` in wp-config.php, for now.

The commit: 
https://github.com/sybrew/the-seo-framework/commit/948e59556f8038d6bc7bd33c7f8832eb086d83bf

**Comment by danielpost** [comment]
@sybrew I don't mean that users wouldn't understand what happens, more that they would be confused there's a button for that in the first place. If you really want to you can add the options but personally I don't think it's a good idea. It's such a trivial thing, I really think you should just pick one and stick to it. Centered is fine.

Another question: are you happy with the current search icon in the WordPress admin menu? I can try making a small logo that has 'TSF' in it or something like that if you want.

**Comment by sybrew** [comment]
Centered it will stay 😄, if any complaints are coming in from end-users (very unlikely) I'll add the option. It's cool to have control sometimes, non-the-less :) 

About the logo: Nope nope nope!
> This plugin is unbranded! This means that we don't even put the name "The SEO Framework" anywhere within the WordPress interface, aside from the plugin activation page. This plugin makes great use of the default WordPress interface elements, like as if this plugin is part of WordPress. No ads, no nags. The small and hidden HTML comment can easily be disabled with the use of a filter or extension.
>
> Nobody has to know about the tools you've used to create your or someone else's website. A clean interface, for everyone.

**Comment by danielpost** [comment]
Ah right, totally forgot about that!

Small fix for the box-shadow on inputs and labels, to fix the blurriness and make the borders a little more subtle:

`.tsf-singular-setting-label,
.tsf-singular-setting-input {   
    box-shadow: inset 0 -1px 0 0 #e9e9e9;
}`

`.tsf-singular-setting-label {
    box-shadow: inset 0 -1px 0 0 #e9e9e9, inset -1px 0 0 0 #e1e1e1;
}`

**Comment by sybrew** [comment]
That change adds unwanted artifacts on mobile view (see the sides at the red lines, they differ from each-other, and also from the input):
image

**Comment by danielpost** [comment]
I see. Giving the second line a min-width (just for desktop) fixes it, at least for me:

screen shot 2017-03-10 at 17 43 03

**Comment by sybrew** [comment]
There's no "just for desktop" in flex 😃 That's the beauty of flexbox :)
Your fix seems to be browser specific. Are you using Webkit or Gecko rather than Blink?

**Comment by danielpost** [comment]
I'm just using the latest version of Chrome. I don't think it's a browser thing, since it's just box-shadow.

The code adds a border at the bottom of each label and input. It also adds a border on the right side of the label, that's what was causing the extra border on the right. Not sure why your inputs didn't get the border, maybe we're using different versions?

**Comment by sybrew** [comment]
I think we're using different commits of this plugin :'). I notice the "Characters used" doesn't have a 15px margin on your screen.

**Comment by danielpost** [comment]
Whoops, my bad. Still though, this should work:

Line 601: change to `box-shadow: inset 0 -1px 0 0 #eee;`
Line 608: add `box-shadow: inset 0 -1px 0 0 #eee, inset -1px 0 0 0 #e1e1e1;`

Add `.tsf-singular-setting-label { box-shadow: inset 0 -1px 0 0 #eee; }` to media-query on line 797.

**Comment by sybrew** [comment]
I believe the exact same can be achieved by only changing line 601.
From: `box-shadow: 0 -1px 1px 0 #e1e1e1;`
To: `box-shadow: 0 0 0 1px #e1e1e1;`

**Comment by danielpost** [comment]
Good catch, although that's a little harsher (I used different colors for the right border and bottom borders and they're all a little softer, and it slightly blurs out the edges of the metabox. This should work:

`box-shadow: -1px 0 0 0 #e1e1e1, 0 0 0 1px #eee;`

Then for the media query, use this: `box-shadow: 0 0 0 1px #eee;`

But either way, it's all very trivial. I'd use #eee or #e9e9e9 over #e1e1e1 though.

**Comment by sybrew** [comment]
I edited my previous comment regarding media-queries 😄

**Comment by danielpost** [comment]
Gotcha, makes sense. I'd say change line 601 to `box-shadow: 0 0 0 1px #e9e9e9;`, that should do it :)

**Comment by sybrew** [comment]
> Also, do you have any plans for the design of the main settings page?

It's quite well accepted, I think... at least from the reviews I read 😄
What do you have in mind?

**Comment by sybrew** [comment]
This is currently so ugly and out of place, but I think I want the tabs at the top for better flow:
image
image

**Comment by danielpost** [comment]
@sybrew Very nice! That's basically what I was thinking, moving the tabs to the top. Is there a specific reason you chose to put the settings in the metabox design? As opposed to standard WordPress settings, e.g. Settings > Reading.

**Comment by sybrew** [comment]
Well, I thought it was neater... The user can move them around, close them when they're done... even easily hide them from view through the Screen Options menu at the top right.
The idea was to use most &mdash; if not all &mdash; available from what WordPress offers.

Most importantly, the POST action options listener wants all options to be included on the page.
A change on how this works is planned for this, however. See #96.

More importantly: Why do you ask?

**Comment by danielpost** [comment]
Fair enough! I was just wondering since it's different from how WordPress itself handles option pages. Since you're using a top-level menu item, you could also show them the same way WordPress shows their options and add subpages for each metabox.

**Comment by sybrew** [comment]
I could, but that's unlikely to happen. For one, it'll be cluttered, and you can't "CTRL+F" the option.
After all, it's only about the SEO optimization part of the site; and nothing else. So it would be most logically placed on one page.

Maybe I could make it basic/advanced... but that would spread out options like "lastmod" from the sitemap metabox. Which will ultimately cause loss of the general flow and structure.

The Title options also directly affect the HomePage options, for example. Try switching the separator, and the options update too. I'm going to add many more of these "feature gimmicks" in the future; that will make everything part of a whole.

All in all, I can't see any logical way for splitting the meta pages.
There's another submenu planned for Extension Options that will use TSFEM's layout, which will essentially be the same as TSF's layout, but then with AJAX and whatnot. And that will be pretty much it regarding the option pages.

The draft:
image

**Comment by sybrew** [comment]
Done and done. 

Could you guys test it and notify me of any issues, or suggestions for improvement?

I've added a few dummy tabs for testing browser resizing.
The collapse and even the column settings should affect all of this.

Also, there's a weird glitch going on when you switch tabs. Where the tab-content container suddenly gets really large. This is only notable when it's the only metabox listed. Also, this is only a frame's worth (1/60) of occurrence, so I'll need to record my screen it to find out why.

**Comment by sybrew** [comment]
It now works on the sidebar metabox holder too!
Flex is truly awesome 😄 

image

Thanks for all your input and for helping kicking off this project @danielpost 👍

**Comment by andremacola** [comment]
Is it possible do remove the box from posts by default?

**Comment by sybrew** [comment]
Hi @andremacola,

Yes, this is possible!
Each post type has its own option for removing metaboxes. This is found in the Screen Options on the top right of your screen.

Alternatively, if you wish to disable this globally, you can do so with this filter:
```
// Remove SEO box globally.
add_filter( 'the_seo_framework_seobox_output', '__return_false' );
```

**Comment by andremacola** [comment]
Hey @sybrew,

Thks, this is what I was looking for. I think you can add in the README.md the Docs link: https://theseoframework.com/docs/api/

I'm translating the plugin to PT-BR by the way.

**Comment by sybrew** [comment]
Hi @andremacola 

Thanks for the translations! 😄 👍 

The filters are found here (one level further):
https://theseoframework.com/docs/api/filters/

The one I mentioned is listed as the 2nd entry :).

Many filters aren't included yet, and some are deprecated or even removed. I still need to rework those pages, but that will take a lot of time I currently am not able to find 🙁.

**Comment by andremacola** [comment]
@sybrew 

I don't think a pull request is necessary just for the translation. Here is the .mo .po files.

https://www.dropbox.com/s/62tkajsta1fsvtn/ptbr.zip?dl=0

**Comment by sybrew** [comment]
Hi @andremacola,

You can import those files right here (at the bottom left if you're logged in):
https://translate.wordpress.org/projects/wp-plugins/autodescription/stable/pt-br/default

<[direct link](https://translate.wordpress.org/projects/wp-plugins/autodescription/stable/pt-br/default/import-translations)>

That way you're also credited for adding those translations 😄. Easy online edits are also available when needed.
Then WordPress will automatically push those translations to all pr_BR users over time.

**Comment by andremacola** [comment]
@sybrew I didn't known that. Thks for the tip. Already imported.

**Pull request merged**